prefetchLanguages.nix

```diff
@@ -0,0 +1,105 @@
+## Because of dynamic loading, Topiary plays poorly in the Nix sandbox. This
+## file introduces two utilities, `prefetchLanguages` and
+## `prefetchLanguagesFile` that transform a Topiary configuration into another
+## one where all the grammars have been pre-fetched and pre-compiled in Nix
+## derivations.
+
+{ pkgs, ... }:
+
+let
+  inherit (builtins)
+    attrNames
+    concatStringsSep
+    mapAttrs
+    toFile
+    readFile
+    toJSON
+    fromJSON
+    baseNameOf
+    ;
+  inherit (pkgs)
+    fetchgit
+    nickel
+    runCommandNoCC
+    writeText
+    tree-sitter
+    ;
+  inherit (pkgs.lib) warn;
+  inherit (pkgs.lib.strings) removeSuffix;
+
+  prefetchLanguageSourceGit =
+    name: source:
+    tree-sitter.buildGrammar {
+      language = name;
+      version = source.rev;
+      src = fetchgit {
+        url = source.git;
+        rev = source.rev;
+        hash =
+          if source ? "nixHash" then
+            source.nixHash
+          else
+            warn "Language `${name}`: no nixHash provided - using dummy value" "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=";
+      };
+      location = if source ? "subdir" then source.subdir else null;
+    };
+
+  prefetchLanguageSource =
+    name: source:
+    if source ? "path" then
+      { inherit (source) path; }
+    else if source ? "git" then
+      { path = "${prefetchLanguageSourceGit name source.git}/parser"; }
+    else
+      throw ("Unsupported Topiary language sources: " ++ concatStringsSep ", " (attrNames source));
```

Comment:
@Xophmeister > Btw, am I right to assume here that the only two types of sources at the moment are `path` and `git`?

Reply:
Yep, that's right :+1: